Port MASTG-TEST-0053 (by @guardsquare) (#3038)
OWASP/owasp-mastg: best-practices/MASTG-BEST-0016.md

---
title: Remove sensitive logs from production builds
alias: remove-logging-in-production
id: MASTG-BEST-0016
platform: ios

> ---

Reviewer: Be careful, only TESTS support `profiles`.

Ensure that all invocations of logging APIs that expose detailed information about the application or user data are removed. Even if a logging API is only active in developer or debug modes, an attacker could:

- Tamper with the device or the app to force it to display all logs, regardless of the build variant
- Gain deeper insight into the application through static analysis by examining verbose log messages

To ensure maximum security, the safest approach is to completely remove these logging calls.

Below is sample code that demonstrates how to eliminate logging APIs from your application.

## 1. Swift

```swift
#if DEBUG
print("Hello world")
#endif
```

> Then you need to set `DEBUG` flag in `Swift Compiler - Custom Flags" > Other Swift Flags` for the development builds.

Reviewer 2: ```suggestion
You need to enable the `DEBUG` flag in `Swift Compiler - Custom Flags > Other Swift Flags` for your development builds.
```

> Then you need to set `DEBUG` flag in `Swift Compiler - Custom Flags" > Other Swift Flags` for the development builds.

Author: I think we look at this issue from a different angle. I think we should merge both approaches. I used a `macro` to make sure that no logging string remains statically in the binary. This would help an attacker to identify the function logic. Let's use your code as the basic requirement and mine as more paranoid.

## 2. Objective-C

```objectivec
#ifdef DEBUG 
# define NSLog (...) NSLog(__VA_ARGS__) 
#else 
# define NSLog (...) 
#endif
```

> Then you need to set `DEBUG` flag in `Apple Clang - Preprocessing > Preprocessor Macros` for the development builds.

Reviewer 2: ```suggestion
You need to enable the `DEBUG` flag in `Apple Clang - Preprocessing > Preprocessor Macros` for your development builds
```

> Then you need to set `DEBUG` flag in `Apple Clang - Preprocessing > Preprocessor Macros` for the development builds.

Reviewer 2: Mention the source for the code  https://stackoverflow.com/questions/6552197/enable-and-disable-nslog-in-debug-mode  in the text

Reviewer: In best practices we need to aim at the best and most up-to-date practices. `NSLog` is old and not recommended anymore, I also don't think that using `print` can be considered a good practice.

Shouldn't we be recommending [OSLog](https://developer.apple.com/documentation/OSLog)/[Logger](https://developer.apple.com/documentation/os/logger) (which rely on Apple's unified logging system available in iOS 10.0 and later) instead? See [Logging](https://developer.apple.com/documentation/os/logging) and https://developer.apple.com/documentation/os/generating-log-messages-from-your-code#Generate-a-Log-Message

Reviewer 2: I have reviewed the document, and indeed, it presents an excellent solution. By default, it automatically conceals dynamic data. We can include the Apple documentation in the best practices.
